docs/Defining_Platforms.md

```diff
@@ -0,0 +1,282 @@
+
+### Support for Buffers of different data types
+
+#### Dynamic allocation and deallocation
+
+This section is TBD:
+
+Two free functions need to be defined in a top-level Buffer.hpp file:
+
+* A factory function templated on the Buffer type:
+```
+template <class BufferT>
+inline BufferT *alloc_buffer(size_t num_elements);
+```
+* A way to deallocate the buffers created by the factory:
+```
+template <class BufferT>
+inline void free_buffer(BufferT *buffer)
+```
+
+TBD: We believe these are needed abstract away how memory is allocated for systems like the Arduino where only static allocation is allowed.
+
+#### Buffer classes for specific types
+
+Untemplated classes are used to declare a different Buffer class for each data type supported by the platform.
+
+The convention used is that FloatBuffer class is declared for floating point types, QUInt8Buffer is used for quantized, unsigned 8-bit integers.
+
+##### Platform constants (again)
+
+Each buffer class must contain static const members for all of the values defined in the corresponding `params_<type>.h` file:
+
+```
+class FloatBuffer
+{
+public:
+    static uint32_t const   W_ob{FLOAT_W_ob};
+    static uint32_t const   C_ob{FLOAT_C_ob};
+    static uint32_t const   SIMD{FLOAT_SIMD};
+    static uint32_t const UNROLL{FLOAT_UNROLL};
+    static uint32_t const   C_ib{FLOAT_C_ib};
+
+    static uint32_t const   NUM_FMA{FLOAT_NUM_FMA};
+    static uint32_t const   NUM_MAX{FLOAT_NUM_MAX};
+    static uint32_t const  NUM_LOAD{FLOAT_NUM_LOAD};
+    static uint32_t const NUM_STORE{FLOAT_NUM_STORE};
+};
+```
+
+Note that the prefix must be removed so that each `<Type>Buffer` class defines exactly the same set of constants.
+
+TBD: make sure these are compile-time constants.
+
+##### Buffer Class typedefs
+
+Two class typedefs must be defined:
+
+```
+class FloatBuffer
+{
+public:
+    typedef float value_type;
+    typedef float accum_type;
+};
+```
+
+*`value_type` - type of data stored in the buffer
+*`accum_type` - mainly used by quantized types to define the type used to accumulate quantized value (usually a larger integer type)
+
+##### Buffer Class Methods API
+
+* `<Type>Buffer()` default constructor - defines a buffer that has no storage, zero elements.
+* `<Type>Buffer(size_t num_elements)` constructor - creates an instance of a buffer with storage for the specified number of elements.
+* Copy constructor
+* Move constructor
+* Copy assignment operator
+* Move assignment operator
+* Destructor (not virtual)
+* `size_t size() const;` - return the number of elements in the buffer
+* `value_type [const] *data() [const];` - return pointer to beginning of buffer memory
+* `value_type [const] &operator[](size_t index) [const];` - index into array using square bracket.
+* `void swap(<Type>Buffer &other);` - swap the contents of this buffer with `other`.
+* `accum_type zero() const;' - mainly used by quantized types to report the bit representation of a zero value.
+
+### Microkernels
```

Comment:
We are going to need more details on how to write the microkernels somewhere

Reply:
Need to get Upasana to document each microkernel.